huaweicloud/services/dns/resource_huaweicloud_dns_endpoint_assignment.go

```diff
@@ -0,0 +1,497 @@
+			"ip_address_id": utils.PathSearch("id", v, nil),
+		}
+
+		for _, subnet := range subnets {
+			if utils.PathSearch("subnet_id", v, "").(string) == utils.PathSearch("neutron_subnet_id", subnet, "").(string) {
+				rst[i]["subnet_id"] = utils.PathSearch("id", subnet, nil)
+				break
+			}
+		}
+	}
+
+	return rst
+}
+
+func resourceEndpointAssignmentUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	conf := meta.(*config.Config)
+	region := conf.GetRegion(d)
+	endpointId := d.Id()
+	client, err := conf.NewServiceClient("dns", region)
+	if err != nil {
+		return diag.Errorf("error creating DNS client: %s", err)
+	}
+
+	if d.HasChange("name") {
+		if err = updateEndpointName(client, endpointId, d.Get("name").(string)); err != nil {
+			return diag.Errorf("error updating the name of the endpoint (%s): %s", endpointId, err)
+		}
+	}
+
+	if d.HasChange("assignments") {
+		err = updateEndpointAssignments(client, d, endpointId)
+		if err != nil {
+			return diag.FromErr(err)
+		}
+
+		stateConf := &resource.StateChangeConf{
+			Target:       []string{"ACTIVE"},
+			Pending:      []string{"PENDING"},
+			Refresh:      refreshEndpointStatus(client, d.Id()),
+			Timeout:      d.Timeout(schema.TimeoutUpdate),
+			Delay:        5 * time.Second,
+			PollInterval: 5 * time.Second,
+		}
+
+		_, err = stateConf.WaitForStateContext(ctx)
+		if err != nil {
+			return diag.Errorf("error waiting for DNS endpoint (%s) update to completed: %s", d.Id(), err)
+		}
+	}
+
+	return resourceEndpointAssignmentRead(ctx, d, meta)
+}
+
+func updateEndpointName(client *golangsdk.ServiceClient, endpointId, name string) error {
+	httpUrl := "v2.1/endpoints/{endpoint_id}"
+	updatePath := client.Endpoint + httpUrl
+	updatePath = strings.ReplaceAll(updatePath, "{endpoint_id}", endpointId)
+	getOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		JSONBody: map[string]interface{}{
+			"name": name,
+		},
+	}
+
+	_, err := client.Request("PUT", updatePath, &getOpt)
+	return err
+}
+
+func updateEndpointAssignments(client *golangsdk.ServiceClient, d *schema.ResourceData, endpointId string) error {
+	var (
+		oldRaws, newRaws = d.GetChange("assignments")
+		addRaws          = newRaws.(*schema.Set).Difference(oldRaws.(*schema.Set)).List()
+		removeRaws       = oldRaws.(*schema.Set).Difference(newRaws.(*schema.Set)).List()
+		// The `originNum` indicates the actual number of IP addresses obtained in the interface.
+		originNum = oldRaws.(*schema.Set).Len()
+		err       error
+	)
+
+	// Since the assignment length ranges from `2` to `6`, the logic is to handle the critical value.
+	// For example, in the following situation:
+	// The original value of assignment is `5`, `2` data items need to be added and `1` data item needs to be deleted.
+	for len(addRaws) > 0 || len(removeRaws) > 0 {
+		if originNum < 6 && len(addRaws) > 0 {
+			addRaws, err = addEndpointIpAddress(client, addRaws, endpointId)
```

Comment:
```suggestion
	// Since the assignment length ranges from `2` to `6`, the logic is to handle the critical value.
	for len(addRaws) > 0 || len(removeRaws) > 0 {
		// If the number of IP address mappings does not reach the upper limit, add an IP address mapping first.
		if originNum < 6 && len(addRaws) > 0 {
			addRaws, err = addEndpointIpAddress(client, addRaws, endpointId)
```